Mathlib/Algebra/Category/ModuleCat/TensorProduct.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` and `(R, S)`-bimodule.
+Consider the tensor functor `(X ⊗[R] .)` from the category of `R`-module to the category of
+`S`-module and the hom functor `X →ₗ[S] .` from the category of `S`-module to the category of
+`R`-module, they form an adjunction. In particular we have that
+```
+Hom_S(X⊗[R]Y, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+-/
+
+universe u u' v
+
+open TensorProduct CategoryTheory
+
+variable (R : Type u) (S : Type u') (X : Type v)
+variable [CommRing R] [CommRing S]
+variable [AddCommGroup X] [Module R X] [Module S X] [SMulCommClass R S X]
+
+namespace ModuleCat
+
+/--
+Let `X` be an `(R,S)`-bimodule, then `(X ⊗[R] .)` is a functor from the category of `R`-modules
+to category of `S`-modules.
+-/
+@[simps!]
+def tensorFunctor : ModuleCat.{v} R ⥤ ModuleCat.{v} S where
+  obj Y := ModuleCat.of S <| X ⊗[R] Y
+  map {Y Y'} l := let L := TensorProduct.map LinearMap.id l
+  { L with
+    map_smul' := fun s x => x.induction_on
+      (show L _ = s • L _ by rw [smul_zero, map_zero, smul_zero])
+      (fun x y => show L _ = s • L _ by rw [smul_tmul', map_tmul, map_tmul, LinearMap.id_apply,
+        smul_tmul', LinearMap.id_apply])
+      (fun x y (hx : L _ = s • L _) (hy : L _ = s • L _) => show L _ = s • L _ by
+        rw [smul_add, map_add, hx, hy, map_add, smul_add]) }
```

Comment:
Is this `TensorProduct.AlgebraTensorModule.map`?

Reply:
I think so, except that `LinearAlgebra.TensorProduct.Tower` assumes some Algebra instances: in my notation it assumes `Algebra R S`, so there is a ring homomorphism which I don't have

Reply:
I think that's probably an indication that the results in `Tower` are in the wrong generality.

Reply:
In `Tower` file, `curry` is a function in the form $M \otimes_R N \to P$ to $M \to_l[A] \to\_l[R] P$ wher $M$ is an $A, R$ bimodule and $N$ an $R$ module and $P$ an $R, A$ bimodule. I think the correct generality is the following:

$M$ is an $R, A$ bimodule and $N$ an $R$ module and $P$ an $A$ module then
$Hom_A(N \otimes[R] M, P) \cong Hom_R(N, Hom_A(M, P))$.

Note that this "flips" `curry` and `uncurry`

so I am not sure if this is what we want in the `Tower` file